Feature/4 콘택트리스트 구현

IamJuco opened the pull request:
close https://github.com/juseonghyun/nbc_ContactApp/issues/4
기능 완성

IamJuco committed:
feat : recyclerview 초기 코드 작성
design : recyclerview 짝수(even)일때, 홀수(odd)일때 아이템및 recyclewview 메인 완성

IamJuco committed:
feat : multi-view 홀수 짝수에 따른 item 설정 완성

IamJuco committed:
feat : inner class -> nested class로 변경, 메모리누수 가능성 + 성능저하 문제 가능성

IamJuco committed:
feat : Recyclerview Item 클릭 시 Detail 데이터전달 및 이동 완성
feat : id("kotlin-parcelize") Plugin 추가및 UserInfo에 적용
feat : AndroidManifest에 DetailActivity추가

IamJuco committed:
feat : Dialog에서 키값으로 데이터 받는 함수 완성
feat : 최종완성

IamJuco committed:
test : 테스트 코드 삭제

IamJuco committed:
Merge remote-tracking branch 'origin/develop' into feature/4-콘택트리스트-구현

IamJuco committed:
comment : 주석으로 코드 설명및 라인정렬

IamJuco committed:
test : 필요없는 테스트 코드 삭제

IamJuco committed:
style : 라인정렬

koreabgj reviewed (approved):
확인했습니다. 수고하셨습니다!

hyezg reviewed (approved):
확인했습니다!

yeonsoo940630 reviewed (approved):
확인했습니다~

yoonchanchoi commented on app/src/main/AndroidManifest.xml:
```diff
@@ -21,6 +21,9 @@
             </intent-filter>
         </activity>
+
+        <activity
```
exported요소가 필요할 것 같습니다.^^

IamJuco replied:
DetailActivity를 말하시는건가요?
DetailActivity exported를 false로 명시해주었습니다!

yoonchanchoi commented on app/build.gradle.kts:
```diff
@@ -1,6 +1,7 @@
 plugins {
     alias(libs.plugins.androidApplication)
     alias(libs.plugins.jetbrainsKotlinAndroid)
+    id("kotlin-parcelize")
```
Parcelize 설정 잘해주셨군요

IamJuco replied:
감사합니다!

IamJuco committed:
style : 라인정렬 + ContactListFragment.kt의 setUpData() 와 SetRecyclerView() 위치 변경

yoonchanchoi commented on app/src/main/java/com/example/nbc_sunnyus/ui/ContactListAdapter.kt:
```diff
@@ -0,0 +1,136 @@
+        notifyItemInserted(position)
+    }
+
+    companion object {
```
compainon object의 코드 컨벤션 위치가 여기가 아닌 거 같아요 😥

https://medium.com/mj-studio/%EC%BD%94%ED%8B%80%EB%A6%B0-%EC%9D%B4%EB%A0%87%EA%B2%8C-%EC%9E%91%EC%84%B1%ED%95%98%EC%8B%9C%EB%A9%B4-%EB%90%A9%EB%8B%88%EB%8B%A4-94871a1fa646
참고 링크입니다.

IamJuco replied:
메서드와 nested class 사이에 compainon object 위치 수정 완료했습니다!

yoonchanchoi commented on app/src/main/java/com/example/nbc_sunnyus/ui/ContactListAdapter.kt:
```diff
@@ -0,0 +1,136 @@
+        const val ODD_VIEW_TYPE = 1 // 홀수 multi-type
+    }
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
```
타입 변수에 따른 layout binding 할당을 잘 해주셧군요😀

IamJuco replied:
감사합니다!

yoonchanchoi commented on app/src/main/java/com/example/nbc_sunnyus/ui/ContactListFragment.kt:
```diff
@@ -1,39 +1,67 @@
 
+    private val userInfo by lazy {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
+            arguments?.getParcelable(Constants.KEY_USER, UserInfo::class.java)
```
데이터 값이 null일 경우에 문제가 생길 것 같습니다. null처리를 해서 null이 아닐 경우에 해주는 시는 건 어떨 까용?

IamJuco replied:
userInfo를 lateinit var로 늦은 초기화해준뒤
setUpData에서 null 체크를 하였습니다.
같이 해결 하였습니다. 감사합니다.

yoonchanchoi commented on app/src/main/java/com/example/nbc_sunnyus/ui/ContactListFragment.kt:
```diff
@@ -1,39 +1,67 @@
 
     private fun setUpListener() {
-        TODO("Not yet implemented")
+        contactListAdapter.setItemClickListener(object : ContactListAdapter.OnItemClickListener {
```
interface를 써서 해당 아이템의 click 이벤트 아이템을 잘 처리 해주셨군요👍
추후에 interface를 밖으로 빼서 캡슐화 해주는 코드를 한번 짜보신는 것도 좋을 것 같습니다.^^

IamJuco replied:
네 알겠습니다!

yoonchanchoi commented on app/src/main/java/com/example/nbc_sunnyus/ui/ContactListFragment.kt:
```diff
@@ -1,39 +1,67 @@
+        })
+    }
+
+    private fun setUpData() {
```
해당 함수명은 데이터를 추가 하는 것과는 맞지 않은 것 같아용😥
setupData보다는 addData() 또는 데이터가 추가 되는 시점에 해당 블럭내부 코드를 쓰는 건 어떨까용?

IamJuco replied:
addData와 setUpData를 분리 후 블럭 내부에서 userInfo를 lazy by 에서 lateinit var로 바꿔주고 setUpData에서 null 처리해줬습니다!

yoonchanchoi commented on app/src/main/java/com/example/nbc_sunnyus/ui/ContactListFragment.kt:
```diff
@@ -1,39 +1,67 @@
+        }
+    }
+
+    private fun setRecyclerView() {
```
어댑터 세팅을 잘해주셨군요 늦은 초기화도 잘 해주셨네용😀

IamJuco replied:
감사합니다!

yoonchanchoi commented on app/src/main/res/layout/activity_main.xml:
```diff
@@ -8,13 +8,10 @@
-        app:layout_constraintEnd_toEndOf="parent"
-        app:layout_constraintStart_toStartOf="parent"
-        app:layout_constraintTop_toTopOf="parent" />
+    <FrameLayout
```
메인 쪽 코드이기 때문에 충돌방지를 위해서 지우는 건 어떨 까용?😥

IamJuco replied:
테스트 코드 삭제완료!

yoonchanchoi commented on app/src/main/res/layout/fragment_contact_list.xml:
```diff
@@ -7,13 +7,14 @@
-        app:layout_constraintTop_toTopOf="parent" />
-
+        app:layout_constraintTop_toTopOf="parent"
+        tools:listitem="@layout/recyclerview_even" />
```
RecyclerView xml을 잘 구현해 주셨네용😀
하지만 정렬하는 것과 listitem을 xml에서 직접 할당하는 것도 좋지만 코드에서 할당한다면 가독성이 더 좋을 거 같아용
listitem은 나중에 한 번 해보시길 ^^

IamJuco replied:
넵! 나중에 한번 시도해보겠습니다!

yoonchanchoi commented on app/src/main/res/layout/fragment_contact_list.xml:
```diff
@@ -7,13 +7,14 @@
+        android:id="@+id/rv_main"
+        android:layout_width="match_parent"
+        android:layout_height="0dp"
+        app:layoutManager="androidx.recyclerview.widget.LinearLayoutManager"
```
layoutManger도 코드에서 한다면 조금 더 많은 기능을 추가 할 수 있을 거 같아요
나중에 한번 해보시길 ㅎㅎ

IamJuco replied:
코드 추가 완료!

yoonchanchoi commented on app/src/main/res/layout/recyclerview_even.xml:
```diff
@@ -0,0 +1,55 @@
+<?xml version="1.0" encoding="utf-8"?>
```
전체적으로 xml 구현 잘해주셨네요👍

IamJuco replied:
감사합니다!

yoonchanchoi commented on app/src/main/res/layout/recyclerview_even.xml:
```diff
@@ -0,0 +1,55 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout
```
정렬 한 번 해주시면 좋을 것 같아요^^

IamJuco replied:
정렬 완료! 사실 이부분은 제가 항상 보기 편하게 엔터 치는데 주의하겠습니다!

yoonchanchoi commented on app/src/main/res/layout/recyclerview_even.xml:
```diff
@@ -0,0 +1,55 @@
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="150dp">
```
wrap_content를 사용해서 높이는 정해서 마진과 패딩을 통해서 뷰 크기를 할당 하는 것은 어떨까요?

IamJuco replied:
wrap_content로 바꿔주고 cardView에서 margin 으로 크기 조정 하였습니다!

yoonchanchoi commented on app/src/main/res/layout/recyclerview_even.xml:
```diff
@@ -0,0 +1,55 @@
+    android:layout_width="match_parent"
+    android:layout_height="150dp">
+
+    <androidx.cardview.widget.CardView
```
카드 뷰로 잘 구현 해주셨네요👍

IamJuco replied:
감사합니다!

yoonchanchoi commented on app/src/main/res/layout/recyclerview_even.xml:
```diff
@@ -0,0 +1,55 @@
+    <View
+        android:layout_width="match_parent"
+        android:layout_height="1dp"
+        android:background="#999999"
```
colors xml에서 함께 관리하는 것은 어떨까요?😥

IamJuco replied:
완료!

yoonchanchoi commented on app/src/main/res/layout/recyclerview_even.xml:
```diff
@@ -0,0 +1,55 @@
+<?xml version="1.0" encoding="utf-8"?>
```
recyclerview_even.xml의 명칭 보다는 item_(아이템요소 Ex : even)로 명칭 하는 건 어떨까용?😥

IamJuco replied:
Item_even
Item_odd
로 둘다 변경완료!

yoonchanchoi commented on app/src/main/res/layout/recyclerview_odd.xml:
```diff
@@ -0,0 +1,55 @@
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="150dp">
```
여기도 wrap_content으로 하시는 건 어떨까용?😅

IamJuco replied:
https://github.com/juseonghyun/nbc_ContactApp/pull/7#discussion_r1578095034
이하동문

yoonchanchoi commented on app/src/main/res/layout/recyclerview_odd.xml:
```diff
@@ -0,0 +1,55 @@
+<?xml version="1.0" encoding="utf-8"?>
```
여기도 item_odd로 명칭을 바꾸는 것은 어떨까용? ^^

IamJuco replied:
https://github.com/juseonghyun/nbc_ContactApp/pull/7#discussion_r1578094483
이하동문

yoonchanchoi commented on app/src/main/res/layout/recyclerview_odd.xml:
```diff
@@ -0,0 +1,55 @@
+        android:id="@+id/tv_phoneNumber"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:text="연락처1"
```
여기도 String xml로 빼거나 tool을 사용하는 것은 어떨까용?

IamJuco replied:
완료!
혹시 예외처리 throw Exception 한 부분도 strings에다가 넣어야할가요?
`else -> throw Exception("Type Error") 여기서 문자열 부분입니다.

yoonchanchoi commented on app/src/main/res/layout/recyclerview_odd.xml:
```diff
@@ -0,0 +1,55 @@
+    <View
+        android:layout_width="match_parent"
+        android:layout_height="1dp"
+        android:background="#999999"
```
colors xml로 빼서 관리하는 건 어떨까용?

IamJuco replied:
https://github.com/juseonghyun/nbc_ContactApp/pull/7#discussion_r1578094132
이하동문

yoonchanchoi reviewed (changes requested):
구현 하시느라 고생하셨습니다. 리뷰 남깁니다. ^^

IamJuco committed:
fix : exported, DetailActivity 추가
style : ContactListAdapter의 compainion object nested class와 메서드 사이로 위치변경
style : ContactListFragment의 setUpData() 함수명을 addData()로 변경
fix : ContactListFragment에서 getParcelable로 받아오던 값이 null일경우 문제 발생, addData()에서 null check
remove : activity_main의 테스트용 코드 삭제
feat : fragment_contact_list.xml에서 LinearLayoutManager를 관리하던것을 ContactListFragment의 setRecyclerview에서 코드로 관리
style : recyclerview_even과 recyclerview_odd의 파일명을 recyclerview -> item으로 변경
style : recyclerview_even과 recyclerview_odd에서 ConstraintLayout 자체에서 높이를 할당하던것을 wrap_content로 기기에 따른 높이 안정성을 추가하고 cardview에서 layout_margin으로 높이 해결
style : color값을 전부 colors.xml로 이동
style : xml의 text를 전부 strings.xml로 이동

IamJuco committed:
fix : userInfo를 lateinit var로 늦은 초기화를 해준 뒤 setUpData()에서 null값 체크

yoonchanchoi reviewed (approved):
수정 다확인했습니다
